In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 

In [2]:
data = pd.read_csv('Titanic-Dataset.csv')

In [3]:
data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
data.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [5]:
data.sample(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
36,37,1,3,"Mamee, Mr. Hanna",male,NaN,0,0,2677,7.2292,NaN,C
406,407,0,3,"Widegren, Mr. Carl/Charles Peter",male,51.0,0,0,347064,7.7500,NaN,S
568,569,0,3,"Doharr, Mr. Tannous",male,NaN,0,0,2686,7.2292,NaN,C
73,74,0,3,"Chronopoulos, Mr. Apostolos",male,26.0,1,0,2680,14.4542,NaN,C


In [6]:
data.shape

(891, 12)

In [7]:
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

In [8]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 118.9 KB


In [9]:
# Checking for null values 
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [10]:
# Summary of numerical column 
data.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [11]:
data.describe(include='object')

/var/folders/wg/8rqwg0nd7995t2tzgq40dsph0000gn/T/ipykernel_17268/3382516499.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  data.describe(include='object')


,Name,Sex,Ticket,Cabin,Embarked
count,891,891,891,204,889
unique,891,2,681,147,3
top,"Braund, Mr. Owen Harris",male,347082,G6,S
freq,1,577,7,4,644


In [12]:
data.duplicated().sum()

np.int64(0)

### Quick Summary 

Total row - 891 

total column(feature) - 12 

Most of the column are numerical Except **Name**, **Sex**, **Ticket**, **Cabin** and **Embarked**

#### Missing value 
There are 177 **Missing value** in the **Age** column, **2 Missing value** in Embarked column and **687 Missing Value** in **Cabin** Column

#### Outliers
**Fare** - The maximum fare of 512 is significantly higher than the average fare of 32.20. 

**Age** - The maximum age of 80 is noticeable higher than the average age of 29.70. Most of the the falls between the 25th percentile (20.13) and the 75th percentile (38.)


In [13]:
data.Name.nunique()/data.shape[0]*100

100.0

In [14]:
data.PassengerId.nunique()/data.shape[0]*100

100.0

In [15]:
data.Ticket.nunique()/data.shape[0]*100

76.43097643097643

In [16]:
data.Ticket.value_counts()

Ticket
347082             7
1601               7
CA. 2343           7
3101295            6
CA 2144            6
                  ..
SOTON/OQ 392076    1
211536             1
112053             1
111369             1
370376             1
Name: count, Length: 681, dtype: int64

In [17]:
data.Cabin.nunique()/data.shape[0]*100

16.4983164983165

In [18]:
data.Cabin.isnull().sum()/data.shape[0]*100

np.float64(77.10437710437711)

Name column is 100% unique and don't contribute in output column so we'll drop this column

Same with PassengerId column 

Same with Ticket column

we'll also drop the **Cabin** column due to a lot of missing value. 

# Handling Missing values 

In [19]:
data.drop(['PassengerId','Name','Ticket', 'Cabin'],axis =1 , inplace=True )

In [20]:
data.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked
0,0,3,male,22.0,1,0,7.2500,S
1,1,1,female,38.0,1,0,71.2833,C
2,1,3,female,26.0,0,0,7.9250,S
3,1,1,female,35.0,1,0,53.1000,S
4,0,3,male,35.0,0,0,8.0500,S


In [21]:
data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [22]:
# Actually i'm in learning phase then i'll apply all the method of filling missing value 


### Imputation

In [23]:
# First fill the missing value with 
# Mean, Median and mode 

In [24]:
medain = data['Age']

In [25]:
median = medain.fillna(medain.mean(), inplace=True)

In [26]:
medain.isnull().sum()

np.int64(0)

In [27]:
mean = data.copy()

In [28]:
mean.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked
0,0,3,male,22.0,1,0,7.2500,S
1,1,1,female,38.0,1,0,71.2833,C
2,1,3,female,26.0,0,0,7.9250,S
3,1,1,female,35.0,1,0,53.1000,S
4,0,3,male,35.0,0,0,8.0500,S


In [29]:
mean['Age'] = mean['Age'].fillna(mean['Age'].median(), inplace=True)

/var/folders/wg/8rqwg0nd7995t2tzgq40dsph0000gn/T/ipykernel_17268/3787048080.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  mean['Age'] = mean['Age'].fillna(mean['Age'].median(), inplace=True)


In [30]:
mean.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [31]:
mean['Embarked'].mode()[0]

'S'

In [32]:
# Let's try mode 
mean['Embarked'] = mean['Embarked'].fillna(mean['Embarked'].mode()[0], inplace=True)

/var/folders/wg/8rqwg0nd7995t2tzgq40dsph0000gn/T/ipykernel_17268/791554111.py:2: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  mean['Embarked'] = mean['Embarked'].fillna(mean['Embarked'].mode()[0], inplace=True)


In [33]:
mean.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [34]:
# We did it using Pandas now it's time to do the same thing using sklearn 
from sklearn.impute import SimpleImputer

from sklearn.impute import SimpleImputer

si_mean = SimpleImputer(strategy="mean")
si_medium = SimpleImputer(strategy='median')
si_mode = SimpleImputer(strategy='most_frequent')
si_constant = SimpleImputer(strategy='7')


In [35]:
si_mean.fit_transform(mean.loc[:, ['Age']])[1:20]

array([[38.],
       [26.],
       [35.],
       [35.],
       [28.],
       [54.],
       [ 2.],
       [27.],
       [14.],
       [ 4.],
       [58.],
       [20.],
       [39.],
       [14.],
       [55.],
       [ 2.],
       [28.],
       [31.],
       [28.]])

In [36]:
si_medium.fit_transform(mean.loc[:, ['Age']])[1:20]

array([[38.],
       [26.],
       [35.],
       [35.],
       [28.],
       [54.],
       [ 2.],
       [27.],
       [14.],
       [ 4.],
       [58.],
       [20.],
       [39.],
       [14.],
       [55.],
       [ 2.],
       [28.],
       [31.],
       [28.]])

In [37]:
mean.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked'],
      dtype='str')

In [38]:
mean['Embarked'] = si_mode.fit_transform(mean.loc[:, ['Embarked']]).ravel()

### Forward/Backward Fill (ffill, bfill)¶

In [44]:
for_data = data.copy()

In [45]:
for_data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [46]:
slicing = for_data.iloc[:7, :4]

In [50]:
slicing

,Survived,Pclass,Sex,Age
0,0,3,male,22.0
1,1,1,female,38.0
2,1,3,female,26.0
3,1,1,female,35.0
4,0,3,male,35.0
5,0,3,male,NaN
6,0,1,male,54.0


In [55]:
slicing['Age'].ffill()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    35.0
6    54.0
Name: Age, dtype: float64

In [56]:
slicing['Age'].bfill()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    54.0
6    54.0
Name: Age, dtype: float64

## Interpolate
Fills missing values using interpolation, which tries to estimate the missing value based on nearby values. 

Best for time series data. 

In [83]:
data4=data.copy()
slicing_4=data4.iloc[:7,:4]
slicing_4

,Survived,Pclass,Sex,Age
0,0,3,male,22.0
1,1,1,female,38.0
2,1,3,female,26.0
3,1,1,female,35.0
4,0,3,male,35.0
5,0,3,male,NaN
6,0,1,male,54.0


In [84]:
slicing_4.drop(columns='Sex', inplace=True)

In [85]:
slicing_4.Age.interpolate(method='linear')

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    44.5
6    54.0
Name: Age, dtype: float64

**interpolation** : Fills missing values using interpolation, which tries to estimate the missing values based on nearby values. 35 + 54 /2 = 44.5

## Drop Missing value 

This method removes rows that contains any missing value. It's useful when you have few missing values and removing them won't significantly impact the dataset. 


dropna()

or drop() to drop a specific column. 

Like we have droped **Cabin** which have around 70% missing values. 

## 4. Useing Replace() for Null Handling

replace() allows you to replace null with a varity of values such as Mean, Median, mode, Constant and unkown values. Also you can replace Null with custom logic. 

In [41]:
null_data = data.copy()
null_data.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [42]:
null_data['Age'] = null_data['Age'].replace(np.nan, null_data['Age'].median(), inplace=True)

/var/folders/wg/8rqwg0nd7995t2tzgq40dsph0000gn/T/ipykernel_17268/4102891513.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  null_data['Age'] = null_data['Age'].replace(np.nan, null_data['Age'].median(), inplace=True)


In [43]:
null_data.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

## KNN Impute 

It's a way to hanle missing values by estimating the missing values based on the values of the nearest neighbors. It leverage the KNN Algorithm to find the most similar data point to the data point with missing values and then imputes the missing value on these neighbours. 

### How it works 
1. Step 1: Identify the data point with missing value 
2. Step 2: Use KNN Algorithm to find the "k" closet data points(neighbors) based on the feature that are not missing. The closeness is usually measured using a distance metric like Euclidean distance. 
3. Step 3: Impute the missing value using the values of the nearest neighbors. There are few ways to do this: 

For continuous variables: Use the mean or weighted average of the neighbors. 

For categorical variables: Use the mode of the nearest neighbors. 

In [72]:
knn_data = data.copy()

In [74]:
slicing_2=knn_data.iloc[:7,:4]
slicing_2

,Survived,Pclass,Sex,Age
0,0,3,male,22.0
1,1,1,female,38.0
2,1,3,female,26.0
3,1,1,female,35.0
4,0,3,male,35.0
5,0,3,male,NaN
6,0,1,male,54.0


In [76]:
slicing_2.drop(columns='Sex', inplace=True)

In [79]:
from sklearn.impute import KNNImputer
knn_imputor = KNNImputer(n_neighbors=2)


# Impute missing value 
imputed_data = knn_imputor.fit_transform(slicing_2)
imputed_data

array([[ 0. ,  3. , 22. ],
       [ 1. ,  1. , 38. ],
       [ 1. ,  3. , 26. ],
       [ 1. ,  1. , 35. ],
       [ 0. ,  3. , 35. ],
       [ 0. ,  3. , 28.5],
       [ 0. ,  1. , 54. ]])

In [82]:
# Convert this into DataFrame 
imputed_data1 = pd.DataFrame(imputed_data, columns=slicing_2.columns)
imputed_data1

,Survived,Pclass,Age
0,0.0,3.0,22.0
1,1.0,1.0,38.0
2,1.0,3.0,26.0
3,1.0,1.0,35.0
4,0.0,3.0,35.0
5,0.0,3.0,28.5
6,0.0,1.0,54.0


## Iterative Impute 
Iterative imputer is a sophisticated method for handling missing values in dataset. It models each feature with missing values as a function of other features and uses an iterative process to fill the missing data. This approach is particularly effective for dataset where the relationship between features can be leverged for more accurate imputation. 

### Key Features of Iterative Imputer. 

- **Model-Based Imputation**
The iterative Impute fits a regression model for each feature with missing values using the other feature in the dataset as predictors. This allows for more nuanced approach compared to similar method like mean or median imputation. 

- **Iterative Process** 
The imputer works in an iterative manner: It starts with by filling in the missing values with an initial guess (like mean imputation) then refine these estimates by fitting the models multiple time until convergence is reached. 

- **Flexable with Different Models:** 
You can specify the estimator used for the imputation process. By default, It uses BayesianRidge regression, but you can use other regressors like Decision Tree, Random Forests, or any other scikit-learn estimator. Handling Multivariate Relationship: 

Since it consider the relationship between multiple features, it can produce more accurate imputations compared to univariate methods. 

### How it works 
- **Initialization:** Missing value are initially filled using a simple imputation method (like mean or median)
- **Model Fitting:**  For each feature with missing values, the imputer fit a regression model using using the other features. 
- **Imputation:** The model predict missing values for that feature. 
- **Itertion:** Step 2 and 3 are repeated for a specified number of iterations or until the predictions stabilize (convergence)

#### Pros and Cons: 
**Pros:** 
More accurate imputations because it takes into account the correlations between features. Can handle different types of features(numerical and categorical)

**Cons:** 
Computationally more intensive compared to simple imputation methods, especially or large dataset. Requires more tuning and understanding of the underlying relationship in the data. 

In [57]:
iter_data = data.copy()

In [64]:
slicing_1 = iter_data.iloc[:7, :4]

In [65]:
slicing_1.drop(columns='Sex', inplace=True)

In [66]:
slicing_1

,Survived,Pclass,Age
0,0,3,22.0
1,1,1,38.0
2,1,3,26.0
3,1,1,35.0
4,0,3,35.0
5,0,3,NaN
6,0,1,54.0


In [67]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# Create an Iterative Imputer instance 
II = IterativeImputer()

In [68]:
# Impute the missing value 
imputed_data = II.fit_transform(slicing_1)


In [69]:
imputed_data

array([[ 0.        ,  3.        , 22.        ],
       [ 1.        ,  1.        , 38.        ],
       [ 1.        ,  3.        , 26.        ],
       [ 1.        ,  1.        , 35.        ],
       [ 0.        ,  3.        , 35.        ],
       [ 0.        ,  3.        , 30.58221364],
       [ 0.        ,  1.        , 54.        ]])

In [71]:
# Convert the result back to a DataFrame 
imputed_data2 = pd.DataFrame(imputed_data, columns=slicing_1.columns) 
imputed_data2

,Survived,Pclass,Age
0,0.0,3.0,22.000000
1,1.0,1.0,38.000000
2,1.0,3.0,26.000000
3,1.0,1.0,35.000000
4,0.0,3.0,35.000000
5,0.0,3.0,30.582214
6,0.0,1.0,54.000000


- **Initialization:** Missing values are initially filled using simple imputation method(Mean, You can specify using initial_strategy )
- **Model Fitting:** For each feature with missing value, the imputer fits a regression model the other features. 
- **Imputation:** The model predict the missing values for that feature. 
- **Iteration:** Step 2 and 3 are repeated for a specified number of iterations until the predictions stabilize(convergence)